api/v1/api.go

```diff
@@ -210,3 +209,71 @@ func (api *API) respondError(w http.ResponseWriter, apiErr apiError, data interf
 		level.Error(api.logger).Log("msg", "failed to write data to connection", "err", err)
 	}
 }
+
+type metricJson map[string]interface{}
+
+func metricsToInterface(metrics []*dto.Metric, metricsType dto.MetricType) []metricJson {
+
+	jsonMetrics := make([]metricJson, len(metrics))
+
+	for i, m := range metrics {
+		individualMetric := make(metricJson)
+		if metricsType == dto.MetricType_SUMMARY {
+			individualMetric["labels"] = makeLabels(m)
+			individualMetric["quantiles"] = makeQuantiles(m)
+			individualMetric["count"] = fmt.Sprint(m.GetSummary().GetSampleCount())
+			individualMetric["sum"] = fmt.Sprint(m.GetSummary().GetSampleSum())
+
+		} else if metricsType == dto.MetricType_HISTOGRAM {
+
+			individualMetric["labels"] = makeLabels(m)
+			individualMetric["buckets"] = makeBuckets(m)
+			individualMetric["count"] = fmt.Sprint(m.GetHistogram().GetSampleCount())
+			individualMetric["sum"] = fmt.Sprint(m.GetHistogram().GetSampleSum())
+		} else {
+
+			individualMetric["labels"] = makeLabels(m)
+			individualMetric["value"] = fmt.Sprint(getValue(m))
+
+		}
+		jsonMetrics[i] = individualMetric
+	}
+	return jsonMetrics
+}
+
+func makeLabels(m *dto.Metric) map[string]string {
+	result := make(map[string]string)
+	for _, lp := range m.Label {
+		result[lp.GetName()] = lp.GetValue()
+	}
+	return result
+}
+
+func makeQuantiles(m *dto.Metric) map[string]string {
+	result := map[string]string{}
+	for _, q := range m.GetSummary().Quantile {
+		result[fmt.Sprint(q.GetQuantile())] = fmt.Sprint(q.GetValue())
+	}
+	return result
+}
+
+func makeBuckets(m *dto.Metric) map[string]string {
+	result := map[string]string{}
+	for _, b := range m.GetHistogram().Bucket {
+		result[fmt.Sprint(b.GetUpperBound())] = fmt.Sprint(b.GetCumulativeCount())
+	}
+	return result
+}
+
+func getValue(m *dto.Metric) float64 {
+	if m.Gauge != nil {
```

Comment:
Again, I think this reads slightly nicer as a switch statement. (See imminent prom2json PR.)